### PR TITLE
[Feat] Pipeline: add overlapped scheduler support

### PR DESCRIPTION
**Pull request by 0xWJ**
Implement overlapped scheduler support by:
1. add cuda stream context for each segment/adapter/reducer
2. add cuda event wait/record for each segment
3. refine sched code generation for stream/event config.

### CODE REVIEW

File: docs/source/parallel_module.md [hunk 1]
```diff
@@ -571,4 +571,37 @@ A common problem with static graph is that it is impossible to handle control fl
...
 
-Please note that user code is flattened and transformed into a single `ParallelModule` at runtime, so `training` is a global module state, and we don't support the case that user want to set a sub-module's training to True but remaining modules to False.
\ No newline at end of file
+Please note that user code is flattened and transformed into a single `ParallelModule` at runtime, so `training` is a global module state, and we don't support the case that user want to set a sub-module's training to True but remaining modules to False.
+
+## Some Details on Integration with Trainer
+
+There are two ways to use `ParallelModule` with Trainer:
+
+1. Pipeline Parallelism with End2End Module (DP/TP can also be used here): You must use `ParallelModule.train_step` and `ParallelModule.infer_step` (which are wrappers of `_train_step`/`_infer_step` from gencode of `ExecutionPlan`) to train/infer the module. The PAS policy must have pipeline parallelism, and the compute config must set `use_end2end=True`.
```
Comment (lynex) [inline docs/source/parallel_module.md]: Abbreviations such as DP, TP, and PP are used without prior explanation. It would be better to use their full expressions to maintain consistency with the previous sections.

Comment (0xWJ) [reply docs/source/parallel_module.md]: fixed.

File: nnscaler/execplan/execplan.py [hunk 2]
```diff
@@ -164,6 +173,33 @@ def block2reuse(node: Block) -> ExeReuseCell:
...
+            if schedplan.stream_config.zero_grad and schedplan.stream_config.zero_grad.stream:
+                return True
+
+            for block in schedplan.nodes():
+                # check segment stream context (all segments are wrapped in blocks)
+                if isinstance(block, Block) and (stream_context := block.stream_context) is not None:
+                    if stream_context.stream:
+                        return True
+                    if stream_context.wait_streams:
+                        return True
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply nnscaler/execplan/execplan.py:194]: by design

File: nnscaler/graph/schedule/schedplan.py [hunk 3]
```diff
@@ -34,9 +34,38 @@
...
+2) Add segments to the plan by `add_segment` or `insert_step` interface.
+3) Call `set_segment_stream` and/or `set_comm_stream_config` to set stream configuration for segments and adapters.
+4) Call `finish` to mark done.
+
 """
 
-from typing import Dict, List,  Optional, Tuple, Set
+from typing import Dict, List,  Optional, Tuple, Set, Union
+from dataclasses import dataclass, field
+from enum import IntEnum
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply nnscaler/graph/schedule/schedplan.py:66]: no fix

File: nnscaler/graph/schedule/schedplan.py [hunk 4]
```diff
@@ -175,6 +257,18 @@ def device(self) -> Tuple[int]:
...
+    @stream_config.setter
+    def stream_config(self, stream_config: StreamConfig):
+        self._stream_config = stream_config
+
+    def _set_node_stream(self, node: IRCell, stream_context: StreamContext):
+        if stream_context is not None:
+            node.set_op_context('stream_context', stream_context)
+
     def nodes(self) -> Tuple[Block]:
         return tuple(self._seqs)
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply nnscaler/graph/schedule/schedplan.py:279]: no fix (legacy code)

File: nnscaler/graph/schedule/schedplan.py [hunk 5]
```diff
@@ -157,6 +235,10 @@ def __init__(self, graph: IRGraph, _dependency: Optional[ScheduleDependency] = N
         self._dependency = _dependency if _dependency is not None \
             else ScheduleDependency(graph)
 
+        # stream context for different adapters,
+        # which can be used in codegen to generate stream context manager for the adapters.
+        self._stream_config: StreamConfig = StreamConfig(cuda_sync_required=False)
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply nnscaler/graph/schedule/schedplan.py:246]: by design

File: tests/cli/test_trainer_overlap.py [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+        dist.all_to_all(out, chunks)
+        return torch.cat(out, dim=0)
+
+    @staticmethod
+    def backward(ctx, grad: torch.Tensor):
+        ws = dist.get_world_size()
+        chunks = [t.contiguous() for t in grad.chunk(ws, dim=0)]
+        out = [torch.empty_like(c) for c in chunks]
+        dist.all_to_all(out, chunks)
+        return torch.cat(out, dim=0), None
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply tests/cli/test_trainer_overlap.py]: fixed.

File: nnscaler/cli/trainer.py [hunk 7]
```diff
@@ -974,7 +974,21 @@ def _train_epoch(self, epoch: int) -> None:
...
             self.hook.on_train_step_start(self, batches[:num_batches])
+
+            # wrap train_step with `torch.cuda.synchronize`
+            # to support multiple cuda streams in train_step.
+            # We never need this for mixed model nor non-pipeline models.
+            cuda_sync_required = isinstance(self.model, nnscaler.ParallelModule) \
+                  and getattr(self.model, 'cuda_sync_required', False)
+
+            if cuda_sync_required:
+                torch.cuda.synchronize()
```
Comment (lynex) [inline nnscaler/cli/trainer.py:985]: This sync is to ensures optimizer.zero_grad(default stream) is fully complete before launching non-default stream communications in train_step, right?

Comment (0xWJ) [reply nnscaler/cli/trainer.py:985]: it will wait all operations in default stream, including dataloader/optimizer step/optimizer zero grad/hooks/.... are done
when we enter train_step, everything is ready.

File: tests/cli/test_trainer_overlap.py [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+    #         torch.cuda.current_stream().wait_stream(nnscaler.runtime.device.DeviceGroup().get_stream('comm'))
+    #         getitem_1_64, linear_69 = nnscaler.runtime.executor.fexecute('InputProj_Segment', model.InputProj_Segment, *(data_57, ), requires_grad=True)
+    #         nnscaler.runtime.device.DeviceGroup().get_event('forward').record()
+
+    #     with torch.cuda.stream(nnscaler.runtime.device.DeviceGroup().get_stream('comm')):
+    #         nnscaler.runtime.device.DeviceGroup().get_event('forward').wait()
+    #         all_to_all_71, getitem_1_109 = nnscaler.runtime.executor.fexecute('CommGather_Segment', model.CommGather_Segment, *(getitem_1_64, linear_69, ), requires_grad=True)
+    #         nnscaler.runtime.device.DeviceGroup().get_event('forward').record()
+
+    #     del getitem_1_64
```
Comment (lynex) [inline tests/cli/test_trainer_overlap.py:350]: How can we ensure the deletion of getitem_1_64 waits until it has been consumed by another stream (line 343), to avoid reading a tensor that no longer exists?

Comment (0xWJ) [reply tests/cli/test_trainer_overlap.py:350]: `del getitem_1_64` only removes a reference.  There is another reference in the GPU queue, so it will not be deallocated. 

I think there is no special here. for example, if you write
```
a = torch.matmul(x, y)
del x
```
when `del x` is executed, `torch.matmul` may haven't start to run (due to the async-ness of cuda ops)

Comment (0xWJ) [reply tests/cli/test_trainer_overlap.py:350]: fixed.

File: docs/source/parallel_module.md [hunk 9]
```diff
@@ -571,4 +571,37 @@ A common problem with static graph is that it is impossible to handle control fl
...
 
-Please note that user code is flattened and transformed into a single `ParallelModule` at runtime, so `training` is a global module state, and we don't support the case that user want to set a sub-module's training to True but remaining modules to False.
\ No newline at end of file
+Please note that user code is flattened and transformed into a single `ParallelModule` at runtime, so `training` is a global module state, and we don't support the case that user want to set a sub-module's training to True but remaining modules to False.
+
+## Some Details on Integration with Trainer
+
+There are two ways to use `ParallelModule` with Trainer:
+
+1. Pipeline Parallelism with End2End Module (Data Parallelism/Tensor Parallelism can also be used here): You must use `ParallelModule.train_step` and `ParallelModule.infer_step` (which are wrappers of `_train_step`/`_infer_step` from gencode of `ExecutionPlan`) to train/infer the module. The PAS policy must have pipeline parallelism, and the compute config must set `use_end2end=True`.
```
Comment (zyeric) [inline docs/source/parallel_module.md:580]: I prefer MPMD for PP and SPMD for Non-PP

File: docs/source/parallel_module.md [hunk 10]
```diff
@@ -571,4 +571,37 @@ A common problem with static graph is that it is impossible to handle control fl
...
+In PP mode, both flags are managed directly in generated code of `ExecutionPlan`, and you don't need to care about them. The codegen will automatically set the flags according to the micro-batch index and the accumulation steps.
+
+In Non-PP mode, If you use `ParallelModule.forward` directly, you need to manually set the flags in the training loop for gradient accumulation by `nnscaler.utils.accum_mode`. If you use `ParellelModule.train_step`, the flags will be automatically set in `train_step` according to the micro-batch index and the accumulation steps, so you don't need to care about them.
+
+
+### Gradient Reduction Support
+
+In the end of `train_step`, we need to sync the gradients across devices. The way we sync gradients is different in PP and Non-PP mode.
+
+We will always call `optimizer.sync_shard_grad()` to sync the gradients before `optimizer.step()`, but in PP mode, the `sync_shard_grad` is a no-op because the gradients are already synced in the codegen, whereas in Non-PP mode, the `sync_shard_grad` will do the real synchronization.
```
Comment (zyeric) [inline docs/source/parallel_module.md]: do we change the implementation? In my understanding, sync_shard_grad is a non-op when the model is in end2end mode.

File: docs/source/parallel_module.md [hunk 11]
```diff
@@ -571,4 +571,37 @@ A common problem with static graph is that it is impossible to handle control fl
...
+
+In Non-PP mode, to support multiple calls of `optimizer.sync_shard_grad()`, `ParallelModule` will keep track whether the gradients are synced or not with `self._sync_grad_required` flag, and only sync the gradients when `self._sync_grad_required` is True. So you can call `optimizer.sync_shard_grad()` multiple times without worrying about it.
+
+We also support async gradient reduction via `compute_config.use_async_reducer`. In this case, the gradient reduction will be kicked off once the gradients are ready, and `optimizer.sync_shard_grad()` will wait for the reduction to be done if it is called before the reduction is done.
+
+When we combine async reduction with gradient accumulation, The time of kicking off gradient reduction becomes a problem. The current implementation is reusing `RuntimeFlag.skip_reducer` flag to control when to kick off the reduction. It is not ideal because `RuntimeFlag.skip_reducer` is originally designed for gradient accumulation, and it is not compatible when overlapping is used. So in overlapped scenarios, we must not use async reduction. We will improve it in the future.
```
Comment (zyeric) [inline docs/source/parallel_module.md:607]: could you please check megatron's implementation? Async reduction is helpful in large scale training, we can refine it in the future.

File: nnscaler/codegen/module/module.py [hunk 12]
```diff
@@ -317,9 +320,11 @@ def scale(self, node: IRCell, device: int) -> IRCell:
             reducer.device = sorted(scale_ranks)
             return reducer
         if isinstance(node, IRSegment) and node.isfw():
+            # TODO: where is the backward segment? Do we need to scale it as well?
```
Comment (zyeric) [inline nnscaler/codegen/module/module.py:323]: this is a trick in my understanding since the backward segment does not need codegen in many cases

Comment (0xWJ) [reply nnscaler/codegen/module/module.py:323]: backward segment is used in Scheduler generation. The current implemenation works, but I will figure it out in future.

File: docs/source/parallel_module.md (hunk 11 above)
Comment (0xWJ) [reply docs/source/parallel_module.md:607]: It is not hard to fix it for overlapping. I will fix it in future.

File: docs/source/parallel_module.md (hunk 10 above)
Comment (0xWJ) [reply docs/source/parallel_module.md]: you are right. fixed.

File: docs/source/parallel_module.md (hunk 9 above)
Comment (0xWJ) [reply docs/source/parallel_module.md:580]: Are they standard terminologies in distributed training? I don't use them because they are not accurate.
for example, MD means DP, but DP is not always used. for TP, I am not sure it is 100% SP. That leads some confusion at least for me.

Comment (zyeric) [reply docs/source/parallel_module.md:580]: In my understanding
MPMD: multiple program multiple data
SPMD: single program multiple data

Comment (0xWJ) [reply docs/source/parallel_module.md:580]: I know that. they are standard terms in computer science. But I am not sure whether they are standard terms in distributed training. I feel these terms are not perfectly matching PP/Non-PP.

File: tests/cli/test_trainer_overlap.py [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+    1F1B overlapped schedule with two CUDA streams.
+
+    Warmup:   full forward  mb 0
+    1F1B:     for i in 1..N-1  →  forward(mb_i) interleaved with backward(mb_{i-1})
+    Cooldown: full backward  mb N-1
+
+    Within each 1F1B step the interleaving pairs forward and backward segments
+    that land on *different* streams so they truly execute concurrently:
+
+        comp_stream: fwd0 │ bwd4 │ fwd2 │ bwd2 │ fwd4 │ bwd0
```
Comment (zyeric) [inline tests/cli/test_trainer_overlap.py:200]: what do fwd2 and bwd3 mean?

File: tests/cli/test_trainer_overlap.py [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+
+    Each stream waits for the other stream's previous event at pair boundaries.
+    """
+
+    if num_microbatches != 4:
+        raise ValueError("This toy schedule is hard-coded for num_microbatches=4")
+    if num_stages != 5:
+        raise ValueError("This toy schedule is hard-coded for num_stages=5")
+
+    segments: List[IRSegment] = graph.select(ntype=IRSegment, flatten=False)
```
Comment (zyeric) [inline tests/cli/test_trainer_overlap.py:213]: when do we assign a stage to a device?

File: tests/cli/test_trainer_overlap.py (hunk 13 above)
Comment (0xWJ) [reply tests/cli/test_trainer_overlap.py:200]: the module forward is split into 5 stages. fwd2 is the second stage, bwd3 is the third stage

File: tests/cli/test_trainer_overlap.py (hunk 14 above)
Comment (0xWJ) [reply tests/cli/test_trainer_overlap.py:213]: i think you are right. It is a potential problem. Currently we follow the logic of autodist in fn policy. 
```
    # stage-wise tensor parallelism
    curr_devices = list(range(ngpus))
    for op_plan in op_plans.values():
        idx = op_plan.stage_id % pp_size
        devs = curr_devices[idx * tp_size: (idx + 1)* tp_size]
        if op_plan.partition is not None:
            _tp(graph, op_plan.op, devs, idx=op_plan.partition.input, dim=op_plan.partition.dim)
        else:
            _replica(graph, op_plan.op, devs)
```

This works for this test case (pp_size=1). For more complex scenarios, we may need a most customizable way. Let's think about it later instead of this PR.

**Review by lynex — APPROVED** [review]
